pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java

```diff
@@ -0,0 +1,53 @@
+package org.pac4j.kerberos.credentials.extractor;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import org.pac4j.kerberos.credentials.KerberosCredentials;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.extractor.CredentialsExtractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.DeferredHttpAction;
+import org.pac4j.core.exception.DeferredHttpActionCallback;
+
+/**
+ * To extract Kerberos headers.
+ *
+ * @author Garry Boyce
+ * @since 1.9.1
+ */
+public class KerberosExtractor implements CredentialsExtractor<KerberosCredentials> {
+    private final String clientName;
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw DeferredHttpAction.deferredHttpAction("Kerberos Header not found", new DeferredHttpActionCallback() {

```

Comment:
I'm a bit worried by the "direct client used in an indirect flow".

Either it's a direct client or an indirect one.

Direct client = credentials are passed in the called url, no redirections, authenticated user (profiles) in request
Indirect client = called url saved and restored, redirections to the identity provider (an url) and back to the application (callback url), authenticated user (profiles) in session

The SecurityFilter protects an url, performs authentication for direct clients and checks that the user is authenticated.

The CallbackFilter receives a callback (indirect client), performs authentication and redirects to the originally called url.

When the authentication is performed in both cases, the `getCredentials` method then the `getUserProfile` method is called: returning null or throwing a BadCredentialsException does not change anything: it means the authentication has failed and no authenticated user is finally available. Returning null means "no credentials was available" while throwing an exception means "based on the given credentials, the authentication failed". But, in both cases, no profile is available and the user is not authenticated.
If you want to break the flow and return a specific header, you need to throw a HttpAction exception.

In the security filter, the authentication **CAN** happen for direct clients while in the callback filter, the authentication **MUST** happen: it means that:
- in the `DirectBasicAuthClient`, in the `BasicAuthCredentialsExtractor`, if no "Basic auth" header exist, `null` or `CredentialsException` are returned -> null credentials -> null profile -> the result is handled by the security filter, maybe a 401 error page if there was only a direct client or a redirection to an identity provider if an indirect client was defined
- in the `IndirectBasicAuthClient`, if no "Basic auth" header exist, an `HttpAction` is sent to force the user to type in its credentials in the basic auth popup.

I need to improve the documentation to better explain this last part. I hope it helps though. Is Kerberos mandatory or optional (coupled with others) as an authentication system?